cdc/cdc.cc

```diff
@@ -578,23 +578,40 @@ static future<::shared_ptr<transformer::streams_type>> get_streams(
     });
 }
 
+template <typename Func>
+future<std::vector<mutation>>
+transform_mutations(std::vector<mutation>& muts, decltype(muts.size()) batch_size, Func&& f) {
+    return parallel_for_each(
+            boost::irange(static_cast<decltype(muts.size())>(0), muts.size(), batch_size),
+            std::move(f))
+        .then([&muts] () mutable { return std::move(muts); });
+}
+
 future<std::vector<mutation>> append_log_mutations(
         db_context ctx,
         schema_ptr s,
         service::storage_proxy::clock_type::time_point timeout,
         service::query_state& qs,
         std::vector<mutation> muts) {
     return get_streams(ctx, s->ks_name(), s->cf_name(), timeout, qs).then([ctx, s = std::move(s), muts = std::move(muts), &qs](::shared_ptr<transformer::streams_type> streams) mutable {
-        return do_with(std::move(muts), transformer(ctx, s, std::move(streams)), [&qs] (std::vector<mutation>& muts, transformer& trans) {
+        return do_with(std::move(muts), transformer(ctx, s, std::move(streams)), [&qs, s] (std::vector<mutation>& muts, transformer& trans) {
             muts.reserve(2 * muts.size());
-            return parallel_for_each(boost::irange(static_cast<decltype(muts.size())>(0), muts.size()), [&qs, &trans, &muts] (int idx) mutable {
-                return trans.pre_image_select(qs.get_client_state(), db::consistency_level::LOCAL_QUORUM, muts[idx]).then([&trans, &muts, idx] (lw_shared_ptr<cql3::untyped_result_set> rs) mutable {
-                    muts.push_back(trans.transform(muts[idx], rs.get()));
+            if (!s->cdc_options().preimage()) {
+                constexpr int batch_size = 100;
+                int muts_len = muts.size();
+                return transform_mutations(muts, batch_size, [&muts, &trans, batch_size, muts_len] (int idx) {
```

Comment:
I don't understand (but I'll still commit....) the whole issue of the "batch size", and why without preimage you use a batch size of 100 (why 100? if you have 100,000,000 mutations this still leaves 1 million concurrent operations), but with preimage you use batch size of 1.

Reply:
This is done here to avoid stalls. pre_image_select was written by Calle and I didn't inspect it for stalls.

Reply:
With a parallel_for_each, there are no stalls. I assume you used "batch size 100" not to avoid stalls but rather, to reduce parallelism, and only then, you switched back to batch size 1 to avoid stalls. All of these measures look very arbitrary to me... A "batch size 100" indeed reduces parallelism, but 100-fold. Why 100-fold? If the original parallelism was 1 million, is 100-fold enough?
There are other ways to do a loop with reasonable parallelism, such as a do_for_each (not parallel) with a semaphore. I'm not saying that the method you used here is completely broken and needs to be fixed immediately, but I am saying it looks pretty weird and arbitrary and it's not clear what it is trying to guarantee and why.

Reply:
I did batches not to reduce the parallelism but to reduce the overhead of future loops.
My reasoning was as follows:
1. I need to break a compute loop to avoid stalls
2. I assume that there's a constant overhead of a iteration of do_for_each or parallel_for_each because of futures and checks so I decided to batch calls to 'transforms' to reduce the impact constant cost of the future loop has on each iteration.

Your remark about the parallelism limitation is very interesting though and worth keeping in mind.